spec/AMQP/20-AMQP_FORMAT.md

```diff
@@ -1,3 +1,121 @@
 # AMQP format
 
-TBD
\ No newline at end of file
+The Advanced Message Queuing Protocol (AMQP) is an open internet protocol for
+business messaging. It defines a binary wire-level protocol that allows for the
+reliable exchange of business messages between two parties. AMQP can be used as
+a protocol for asynchronous communication between components of an application.
+From the distributed tracing and telemetry correlation perspective it is a
+known problem to be able to correlate a component that placed message and
+component that processed it later. This specification describes how trace
+context MUST be encoded into AMQP messages.
+
+## Trace context fields placement in a message
+
+AMQP defines message as a payload with the additional annotations sections.
+There are two annotation sections this specification refers -
+"application-properties" and "message-annotations". See
+[3.2](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#section-message-format)
+of *OASIS Advanced Message Queuing Protocol (AMQP) Version 1.0, Part 3:
+Messaging* (AMQP specification).
+
+AMQP message section "application-properties" is immutable collection of
+properties that is defined by message publisher and can be read by the message
+consumer. Message brokers cannot mutate those properties. Section
+"message-annotations" is designed for message brokers to use and can be mutated
+during the message processing.
+
+Fields `traceparent` and `tracestate` SHOULD be added to the message in the
+application-properties section by message publisher. Once the message has been
+created, it’s no longer permissible to edit the bare message. So if it were
+necessary to annotate the message inside the middleware as it transits, that
+MUST happen in the “message-annotations” section, using the
+“application-properties” as a base. Message reader SHOULD construct the full
+trace context by reading `traceparent` and `tracestate` fields from the
+“message-annotations” first and if not exist - from “application-properties”.
```

Comment:
We have been looking into where to put tracing context information in a message as well (in the context of adding OpenTracing support to Eclipse Hono and to EnMasse) and have been discussing whether to use the "message-annotations" section or the "delivery-annotations" for that. We chose the "delivery-annotations" - see the discussion here: EnMasseProject/enmasse#1212.

One point here was about the part of the AMQP spec where it is said that "message annotations SHOULD be propagated across every delivery step".
There might be cases where such a requirement looks useful, but in other cases it might be desirable for a component to not forward tracing info any further - to not propagate tracing info to _external_ application level message consumers.

Have you considered using delivery-annotations instead of message-annotations so that the propagation of tracing context information can be defined on a "per hop" basis? 
Thereby tracing context information could either be directly associated with the message (via the immutable application-properties) or  - to override this - via delivery-annotations with the flexibility of either propagating it further or not.

Reply:
My understanding is that "message-annotation" are overridable. So you getting benefit of them being propagated, but still can edit them or remove them when needed. Do you know what other isues "message-annotations" may bring? Or my understanding of "message-annotations" is incorrect?

@clemensv I'd appreciate if you can chime in here.

Reply:
Several thoughts on this. 

First, I like the integration as it stands and I think this is good overall. 

On the application-properties vs. message-annotations vs. delivery-annotations question, I'm where @rgodfrey stood on the issue that @calohmn quoted: https://github.com/EnMasseProject/enmasse/issues/1212#issuecomment-396868699

I think manipulations of the context made during routing ought to go into the message annotations.

I think there are three grand scenarios:

- Application trace context through a simple AMQP brokered entity. If you send a message into a queue from an app and then pull that message from the queue with the same or another app and nothing else interesting happens, application-properties is the right place, because you're not doing any tracing of the broker, you're just passing the context alongside the message.

- Application context plus routing context. If you have a chain of AMQP entities as you might have in a complex routing scenario, you might want to enrich the context as the message is being passed on. In that case, the prior scenario still exists, i.e. you will still want the option to have the context information as set by the message producer and use that for application level tracing unencumbered by whatever the routing layer might be tracking. As the routing layer gets it hand on the message for tracing, it basically forks the context off into the message-annotations and that's being manipulated on the route. As the message reaches the destination, you now have the app-level context from the producer in the application-properties AND the routing context that was layered on top of that in the message-annotations.

- Hop to hop tracing: A delivery-annotations usage scenario would be purely additive to either of the prior two, allowing any intermediary to spawn a new context for itself or to propagate a context only via subset of hops. I think that is fairly esoteric at this point and I wouldn't take that into a spec unless someone shows a hard use-case.

Reply:
Added verbatim to rationale document and resolving this conversation. @calohmn feel free to re-open